File: app/src/main/java/com/abhi/lokaldemo/ui/main/adapter/PostListAdapter.java
package com.abhi.lokaldemo.ui.main.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.View.OnClickListener;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.abhi.lokaldemo.R;
import com.abhi.lokaldemo.api.model.PostResponse;
import com.abhi.lokaldemo.databinding.PostListItemBinding;
import com.abhi.lokaldemo.ui.main.callback.PostClickCallback;
import com.abhi.lokaldemo.ui.main.callback.PostClickWithBindingCallback;

import java.util.List;
import java.util.Objects;

public class PostListAdapter extends RecyclerView.Adapter<PostListAdapter.PostViewHolder> {


    List<? extends PostResponse> postList;

    @Nullable
    private final PostClickWithBindingCallback postClickCallback;

    public PostListAdapter(@Nullable PostClickWithBindingCallback postClickCallback) {
        this.postClickCallback = postClickCallback;
    }


    public void setProjectList(final List<? extends PostResponse> postList) {
        if (this.postList == null) {
            this.postList = postList;
            notifyItemRangeInserted(0, postList.size());
        } else {
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
                @Override
                public int getOldListSize() {
                    return PostListAdapter.this.postList.size();
                }

                @Override
                public int getNewListSize() {
                    return postList.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return PostListAdapter.this.postList.get(oldItemPosition).id ==
                            postList.get(newItemPosition).id;
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    PostResponse newPost = postList.get(newItemPosition);
                    PostResponse oldPost = postList.get(oldItemPosition);
                    return newPost.id == oldPost.id;
                }
            });
            this.postList = postList;
            result.dispatchUpdatesTo(this);
        }
    }


    @NonNull
    @Override
    public PostViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final PostListItemBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.post_list_item,
                        parent, false);

        binding.setCallback(new PostClickCallback() {
            @Override
            public void onClick(PostResponse post) {
                postClickCallback.onClick(post, binding);
            }
        });

        return new PostViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull PostViewHolder holder, int position) {
        holder.binding.setPost(postList.get(position));
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return postList.size();
    }


    static class PostViewHolder extends RecyclerView.ViewHolder {

        final PostListItemBinding binding;

        public PostViewHolder(PostListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/abhi/lokaldemo/ui/main/adapter/PostListAdapter.kt
package com.abhi.lokaldemo.ui.main.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.abhi.lokaldemo.R
import com.abhi.lokaldemo.api.model.PostResponse
import com.abhi.lokaldemo.databinding.PostListItemBinding
import com.abhi.lokaldemo.ui.main.callback.PostClickCallback


class PostListAdapter(
    val projectPostClickCallback: PostClickCallback
) : RecyclerView.Adapter<PostListAdapter.PostViewHolder>() {

    var postList: List<PostResponse>? = null

    fun setProjectList(postList: List<PostResponse>) {
        if (this.postList == null) {
            this.postList = postList
            notifyItemRangeInserted(0, postList.size)
        } else {
            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return postList.size
                }

                override fun getNewListSize(): Int {
                    return postList.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return postList[oldItemPosition].id == postList[newItemPosition].id
                }

                override fun areContentsTheSame(
                    oldItemPosition: Int,
                    newItemPosition: Int
                ): Boolean {
                    val post = postList[newItemPosition]
                    val old = postList[oldItemPosition]
                    return post.id == old.id
                }
            })
            this.postList = postList
            result.dispatchUpdatesTo(this)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PostViewHolder {
        val binding: PostListItemBinding = DataBindingUtil
            .inflate(
                LayoutInflater.from(parent.context), R.layout.post_list_item,
                parent, false
            )

        binding.callback = projectPostClickCallback

        return PostViewHolder(binding)
    }

    override fun onBindViewHolder(holder: PostViewHolder, position: Int) {
        holder.binding.post = postList!![position]
        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        return postList?.size ?: 0
    }

    class PostViewHolder(val binding: PostListItemBinding) :
        RecyclerView.ViewHolder(binding.getRoot())
}
